shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java

```diff
@@ -1,178 +1,275 @@
     if (isApplicationRunning) {
-      throw new AssertionError("This Flutter engine instance is already running an application");
+      Log.w(TAG, "Attempted to run a DartExecutor that is already running.");
+      return;
     }
 
     flutterJNI.nativeRunBundleAndSnapshotFromLibrary(
         nativeObjectReference,
-        bundlePath,
-        defaultPath,
-        entrypoint,
-        libraryPath,
-        resources.getAssets()
+        dartCallback.pathToPrimaryBundle,
+        dartCallback.pathToFallbackBundle,
+        dartCallback.callbackHandle.callbackName,
+        dartCallback.callbackHandle.callbackLibraryPath,
+        dartCallback.androidAssetManager
     );
 
     isApplicationRunning = true;
   }
 
-  public void stop() {
-    isApplicationRunning = false;
+  //------ START BinaryMessenger -----
+  /**
+   * Sends the given {@code message} from Android to Dart over the given {@code channel}.
+   *
+   * @param channel the name of the logical channel used for the message.
+   * @param message the message payload, a direct-allocated {@link ByteBuffer} with the message bytes
+   */
+  @Override
+  public void send(String channel, ByteBuffer message) {
+    messenger.send(channel, message, null);
   }
 
-  // TODO(mattcarroll): Implement observatory lookup for "flutter attach"
-  public String getObservatoryUrl() {
-    return flutterJNI.nativeGetObservatoryUri();
+  /**
+   * Sends the given {@code messages} from Android to Dart over the given {@code channel} and
+   * then has the provided {@code callback} invoked when the Dart side responds.
+   *
+   * @param channel the name of the logical channel used for the message.
+   * @param message the message payload, a direct-allocated {@link ByteBuffer} with the message bytes
+   * between position zero and current position, or null.
+   * @param callback a callback invoked when the Dart application responds to the message
+   */
+  @Override
+  public void send(String channel, ByteBuffer message, BinaryMessenger.BinaryReply callback) {
+    messenger.send(channel, message, callback);
   }
 
+  /**
+   * Sets the given {@link io.flutter.plugin.common.BinaryMessenger.BinaryMessageHandler} as the
+   * singular handler for all incoming messages received from the Dart side of this Dart execution
+   * context.
+   *
+   * @param channel the name of the channel.
+   * @param handler a {@link BinaryMessageHandler} to be invoked on incoming messages, or null.
+   */
   @Override
   public void setMessageHandler(String channel, BinaryMessenger.BinaryMessageHandler handler) {
-    if (handler == null) {
-      messageHandlers.remove(channel);
-    } else {
-      messageHandlers.put(channel, handler);
-    }
+    messenger.setMessageHandler(channel, handler);
   }
+  //------ END BinaryMessenger -----
 
-  @Override
-  public void send(String channel, ByteBuffer message) {
-    send(channel, message, null);
+  // TODO(mattcarroll): Implement observatory lookup for "flutter attach"
```

Comment:
I'm assuming this should be covered by the zeroconf work for iOS and equivalent work on the Android side?

Reply:
What is the zeroconf work?